rfcs/2021-08-06-8619-framing-and-codecs-sources.md

```diff
@@ -0,0 +1,212 @@
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and decoding, such that each source may implement these in their own
+way. This has led to inconsistencies in available options and poses additional
+maintenance burden. Components have their own defaults, making
+behavior unpredictable and surprising users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+To expose framing and decoding to sources in a uniform way, we specify a common
+configuration struct:
+
+```rust
+pub struct DecodingConfig {
+    framing: Option<FramingConfig>,
+    decoding: Option<ParserConfig>,
+}
+```
+
+while the framing method and parser used to deserialize into a structured event
+can be selected from an enum:
+
+```rust
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum FramingConfig {
+    NewlineDelimited,
+    CharacterDelimited {
+        delimiter: char,
+        max_length: Option<usize>,
+    },
+    OctetCounting {
+        max_length: Option<usize>,
+    },
+    ...
+}
+
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ParserConfig {
+    Bytes,
+    Json {
+        ...
+    },
+    Syslog,
+    ...
+}
+```
+
+The `DecodingConfig` exposes a `build` method to create a Vector `Decoder` which
+implements the `tokio_util::codec::Decoder<Item = (Event, usize)>` trait. That
+way we can produce `Event`s from `ByteMut` by repeated calls to
+`decode`/`decode_eof` either on a byte stream or a byte message. The additional
+`usize` item conveys how many bytes were read to produce the particular `Event`,
+so that this information can be passed along to e.g. the internal event log.
+
+Internally, the `Decoder` holds a framer and a parser:
+
+```rust
+pub struct Decoder {
+    framer: Box<dyn tokio_util::codec::Decoder<Item = Bytes> + Send + Sync>,
+    parser: Box<dyn Parser + Send + Sync>,
+}
+```
+
+with the `Parser` trait being defined as:
+
+```rust
+pub trait Parser {
+    fn parse(&self, bytes: Bytes) -> crate::Result<Event>;
+}
+```
+
+Consequently, the `FramingConfig` exposes a `build` method to create a
+`Box<dyn tokio_util::codec::Decoder<Item = Bytes> + Send + Sync>` and the
+`ParserConfig` to create a `Box<dyn Parser + Send + Sync>` which are placed into
+the `Decoder`.
+
+The `Decoder` calls its framer repeatedly to produce byte frames, then calls the
+parser to create an `Event` and returns.
+
+Sources which want to expose framing/decoding functionality to the user can
+embed `DecodingConfig` in their config, build the `Decoder` and apply it to the
+sequences of bytes they produce to create events.
+
+## Doc-level Proposal
+
+The `framing` and `decoding` behavior can be configured for a source in the
+vector config, e.g.:
+
+```yaml
+[framing]
+type = "character_delimited"
+delimiter = "\t"
+max_length = 1337
+
+[decoding]
+type = "json"
```

Comment:
On the sink side of things, we call this `codec` rather than `type`. Would it make sense to use the same terminology here?

Reply:
Certainly! 😄 What do you think about the `framing` key. Should we leave it at `type` or change it to e.g. `framer` or `codec`?

Reply:
Good question! I think `type` is fine since `framing` is single-purpose. `encoding` does multiple things: handles encoding of events (`codec`), excluding fields (`except_fields`), limiting to fields (`only_fields`) so I think `codec` makes sense there rather than `type` since the `codec` is unrelated to the other options which are applied independently.

However, this is also making me think about the modeling of the config options a bit more. Starting with the `aws_s3` source, we've started to nest context dependent options (for example options that only apply if `type`, `strategy`, etc. is some value) rather than having top-level keys. In my opinion, this makes it a bit clearer when which options apply.

In your example, I think this would make the framing config like:

```
[framing]
type = "character_delimited"
charater_delimiter.delimiter = "\t"
max_length = 1337
```

I think `max_length` can stay at the top since it should apply to all framers.

Reply:
Adjusted in 9b8332bf8bf377f8c6b185c69b6ee42325d91e6a